dotnet/src/SemanticKernel.Abstractions/Orchestration/FunctionResult.cs

```diff
@@ -127,5 +125,78 @@ public bool TryGetMetadataValue<T>(string key, out T value)
     }
 
     /// <inheritdoc/>
-    public override string ToString() => this.Value?.ToString() ?? base.ToString();
+    public override string ToString()
+    {
+        var stringResult = ConvertToString(this.Value, this.Culture);
+
+        return string.IsNullOrEmpty(stringResult) ? string.Empty : stringResult!;
+    }
+
+    private static string? ConvertToString(object? value, CultureInfo culture)
+    {
+        if (value == null) { return null; }
+
+        var sourceType = value.GetType();
+
+        var converterFunction = GetTypeConverterDelegate(sourceType);
+
+        return converterFunction == null
+            ? value.ToString()
+            : converterFunction(value, culture);
+    }
+
+    private static Func<object?, CultureInfo, string?>? GetTypeConverterDelegate(Type sourceType) =>
+        s_converters.GetOrAdd(sourceType, static sourceType =>
+        {
+            // Strings just render as themselves.
+            if (sourceType == typeof(string))
+            {
+                return (input, cultureInfo) => (string)input!;
+            }
+
+            // Look up and use a type converter.
+            if (GetTypeConverter(sourceType) is TypeConverter converter && converter.CanConvertTo(typeof(string)))
+            {
+                return (input, cultureInfo) =>
+                {
+                    return converter.ConvertToString(context: null, cultureInfo, input);
+                };
+            }
+
+            return null;
+        });
+
+    private static TypeConverter? GetTypeConverter(Type sourceType)
```

Comment:
Can we consolidate this into a shared file used by both this code and the marshaling logic in KernelFunctionFromMethod?

Reply:
Yep, absolutely. It will be addressed in scope of this issue - https://github.com/microsoft/semantic-kernel/issues/3761 shortly after the PR is merged.